pkg/kademlia/queue.go

```diff
@@ -4,57 +4,99 @@
 package kademlia
 
 import (
+	"container/heap"
 	"math/big"
 
+	"storj.io/storj/pkg/dht"
 	"storj.io/storj/pkg/pb"
 )
 
-// An Item is something we manage in a priority queue.
-type Item struct {
+//XorQueue is a priority queue where the priority is key XOR distance
+type XorQueue struct {
+	maxLen int
+	items  items
+}
+
+//NewXorQueue returns a items with priority based on XOR from targetBytes
+func NewXorQueue(size int) *XorQueue {
+	return &XorQueue{items: make(items, 0, size), maxLen: size}
+}
+
+//Insert adds Node onto the queue
+func (x *XorQueue) Insert(target dht.NodeID, nodes []*pb.Node) {
+	targetBytes := new(big.Int).SetBytes(target.Bytes())
+	//insert new nodes
+	for _, node := range nodes {
+		heap.Push(&x.items, &item{
+			value:    node,
+			priority: new(big.Int).Xor(targetBytes, new(big.Int).SetBytes([]byte(node.GetId()))),
+		})
+	}
+	//resize down if we grew too big
+	if x.items.Len() > x.maxLen {
+		olditems := x.items
+		x.items = items{}
+		for i := 0; i < x.maxLen && len(olditems) > 0; i++ {
+			item := heap.Pop(&olditems)
+			heap.Push(&x.items, item)
+		}
+		heap.Init(&x.items)
+	}
+}
+
+//Closest removed the closest priority node from the queue
+func (x *XorQueue) Closest() (*pb.Node, dht.NodeID) {
+	item := *(heap.Pop(&x.items).(*item))
+	return item.value, item.priority
```

Comment:
Priority isn't the nodeID though but the XOR distance from the nodeID and the target

Reply:
Ok, this should probably explicitly return a big.Int, rather than the dht.NodeID interface.  We already have calculated the Xor before the priority get stored, so this should mean just changing the type.